pkg/receive/config.go

```diff
@@ -0,0 +1,206 @@
+package receive
+
+import (
+	"context"
+	"encoding/json"
+	"io/ioutil"
+	"os"
+	"reflect"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/common/model"
+	"gopkg.in/fsnotify.v1"
+)
+
+// HashringConfig represents the configuration for a hashring
+// a receive node knows about.
+type HashringConfig struct {
+	Hashring  string   `json:"hashring"`
+	Tenants   []string `json:"tenants"`
+	Endpoints []string `json:"endpoints"`
+}
+
+// ConfigWatcher is able to watch a file containing a hashring configuration
+// for updates.
+type ConfigWatcher struct {
+	path     string
+	interval time.Duration
+	logger   log.Logger
+	watcher  *fsnotify.Watcher
+
+	changesCounter prometheus.Counter
+	errorCounter   prometheus.Counter
+	refreshCounter prometheus.Counter
+
+	// last is the last known configuration.
+	last []HashringConfig
+}
+
+// NewConfigWatcher creates a new ConfigWatcher.
+func NewConfigWatcher(path string, interval model.Duration, logger log.Logger) (*ConfigWatcher, error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, errors.Wrap(err, "creating file watcher")
+	}
+	if err := watcher.Add(path); err != nil {
+		return nil, errors.Wrap(err, "adding path to file watcher")
+	}
+	return &ConfigWatcher{
+		path:     path,
+		interval: time.Duration(interval),
+		logger:   logger,
+		watcher:  watcher,
+		changesCounter: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "thanos_receive_hashrings_file_changes_total",
+				Help: "The number of times the hashrings configuration file has changed.",
+			}),
+		errorCounter: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "thanos_receive_hashrings_file_errors_total",
+				Help: "The number of errors watching the hashrings configuration file.",
+			}),
+		refreshCounter: prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "thanos_receive_hashrings_file_refreshes_total",
+				Help: "The number of refreshes of the hashrings configuration file.",
+			}),
+	}, nil
+}
+
+// RegisterMetrics registers the configuration watcher's metrics
+// with the given Prometheus registerer.
+func (cw *ConfigWatcher) RegisterMetrics(r prometheus.Registerer) {
```

Comment:
this is typically done in the `XYZNew` function and if `prometheus.Registerer` is nil, then it's just not, see here for example: https://github.com/improbable-eng/thanos/blob/e1d3b6297f194879d9f0653abb0479e30ffc2163/pkg/shipper/shipper.go#L110